### PR TITLE
Smarter schema visitor

### PR DESCRIPTION
**Pull request by bbakerman**
In short I want to schema visitors to easier to use.  eg if I am on a field def, can I get its containing type.  Not `context.getParentObject()` and cast it but a type safe mechanism for that.

This is more like the `graphql.analysis.QueryVisitor#visitField` that exists which gives out  stuff like

```
default TraversalControl visitFieldWithControl(QueryVisitorFieldEnvironment queryVisitorFieldEnvironment) {
        visitField(queryVisitorFieldEnvironment);
        return TraversalControl.CONTINUE;
    }

    void visitInlineFragment(QueryVisitorInlineFragmentEnvironment queryVisitorInlineFragmentEnvironment);

    void visitFragmentSpread(QueryVisitorFragmentSpreadEnvironment queryVisitorFragmentSpreadEnvironment);

    default void visitFragmentDefinition(QueryVisitorFragmentDefinitionEnvironment queryVisitorFragmentDefinitionEnvironment) {

    }
```

So this has now progressed and I have more environments available - all of the major types plus one catch all `GraphqlSchemaElement` one

I dont visit Non nulls or Lists - I dont think this is useful 

If you really want to do that - then use a plain old TypeVisitor - all the power without the nice ness

### CODE REVIEW

File: src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    default TraversalControl visitGraphQLObjectType(GraphQLObjectType objectType, SmartTypeVisitorEnvironment environment) {
+        return TraversalControl.CONTINUE;
+    }
+
+    /**
+     * This is a class specific for visiting {@link GraphQLFieldDefinition}s
+     */
+    interface GraphQLFieldDefinitionVisitorEnvironment extends SmartTypeVisitorEnvironment {
+
+        GraphQLFieldsContainer getFieldsContainer();
```
Comment (bbakerman) [inline src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java:27]: One other option here is to NOT have a per method interface but rather put this into method signature

```
visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, GraphQLFieldsContainer 
 container, SmartTypeVisitorEnvironment environment) {
```

Comment (bbakerman) [reply src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java:27]: What else would go here to make it smarter ?

File: src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * @param fieldDefinition the field
+     * @param environment     the specific env
+     *
+     * @return control
+     */
+    default TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, GraphQLFieldDefinitionVisitorEnvironment environment) {
+        return TraversalControl.CONTINUE;
+    }
+
+
```
Comment (bbakerman) [inline src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java:43]: Clearly I need to add all the other visi methods - eg arguments, directives, unions and so on

File: src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java (hunk 1 above)
Comment (felipe-gdr) [reply src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java:27]: What about some callback methods for the intermediate interfaces in the class hierarchy?
Like, I often find myself having to do stuff on all elements that can have directives. 
Maybe a callback like `visitGraphQLDirectiveContainer(GraphQLDirectiveContainer directiveContainer.....)`? `GraphQLFieldsContainer` is another very common one.
Although that can quickly become convoluted - do we need one for each of those intermediate interfaces (`GraphQLNamedInputType`, `GraphQLNullableType`, etc...)? Probably not.

Comment (felipe-gdr) [reply src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java:27]: maybe like a dynamic visitor method`<T extends GraphQLSchemaElement> visitThing(Class<T> thingClass, BiConsumer<T, Context>)` , and in the visitor we could check `thingClass.isAssignableFrom(element.getClass())`. Something like that...
Not sure if there's already something like that in the API, though.

File: src/main/java/graphql/schema/GraphQLTypeVisitor.java [hunk 3]
```diff
@@ -13,54 +13,56 @@
...
-     * and this is another place that has been left in.  In the future this behavior will change and this will only visit directive definitions of a schema, not where
-     * they are applied.
      *
-     * @param node the directive
+     * @param node    the applied directive
      * @param context the traversal context
+     *
      * @return how to control the visitation processing
      */
-    TraversalControl visitGraphQLDirective(GraphQLDirective node, TraverserContext<GraphQLSchemaElement> context);
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLTypeVisitor.java:39]: I ordered all the methods here in alphabetic order so I can know what's what

File: src/main/java/graphql/schema/SchemaTransformer.java [hunk 4]
```diff
@@ -271,6 +271,9 @@ public TraversalControl backRef(TraverserContext<GraphQLSchemaElement> context)
         if (codeRegistry != null) {
             traverser.rootVar(GraphQLCodeRegistry.Builder.class, codeRegistry);
         }
+        if (schema != null) {
+            traverser.rootVar(GraphQLSchema.class, schema);
+        }
```
Comment (bbakerman) [inline src/main/java/graphql/schema/SchemaTransformer.java:276]: We now add the source schema as a root variable - which visitors can now access

File: src/main/java/graphql/schema/SchemaTraverser.java [hunk 5]
```diff
@@ -66,9 +66,11 @@ public TraverserResult depthFirstFullSchema(List<GraphQLTypeVisitor> typeVisitor
         roots.addAll(schema.getAdditionalTypes());
         roots.addAll(schema.getDirectives());
         roots.addAll(schema.getSchemaDirectives());
+        roots.addAll(schema.getSchemaAppliedDirectives());
         roots.add(schema.getIntrospectionSchemaType());
         TraverserDelegateListVisitor traverserDelegateListVisitor = new TraverserDelegateListVisitor(typeVisitors);
-        return initTraverser().rootVars(rootVars).traverse(roots, traverserDelegateListVisitor);
+        Traverser<GraphQLSchemaElement> traverser = initTraverser().rootVars(rootVars).rootVar(GraphQLSchema.class, schema);
+        return traverser.traverse(roots, traverserDelegateListVisitor);
```
Comment (bbakerman) [inline src/main/java/graphql/schema/SchemaTraverser.java:73]: Same here on SchemaTraverser - the schema being traversed is now a root variable

File: src/main/java/graphql/schema/SchemaTraverser.java [hunk 6]
```diff
@@ -131,7 +133,10 @@ private static class TraverserDelegateListVisitor implements TraverserVisitor<Gr
         @Override
         public TraversalControl enter(TraverserContext<GraphQLSchemaElement> context) {
             for (GraphQLTypeVisitor graphQLTypeVisitor : typeVisitors) {
-                context.thisNode().accept(context, graphQLTypeVisitor);
+                TraversalControl control = context.thisNode().accept(context, graphQLTypeVisitor);
+                if (control != CONTINUE) {
+                    return control;
+                }
```
Comment (bbakerman) [inline src/main/java/graphql/schema/SchemaTraverser.java:139]: This was broken.  I fixed it and now tests broke

So I wrote a new test for this case - when NOT CONTINUE is returned

File: src/main/java/graphql/schema/visitor/GraphQLSchemaTraversalControl.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
+package graphql.schema.visitor;
```
Comment (bbakerman) [inline src/main/java/graphql/schema/visitor/GraphQLSchemaTraversalControl.java:1]: We have a different "control" that SchemaTraversalControl - so we can have better methods and "hide" the `changeNode` malarky we must do.

The only way to get a GraphQLSchemaTraversalControl (which is package level on creation) is by the helper methods and hence this simplifies the call API for consumers

File: src/main/java/graphql/schema/visitor/GraphQLSchemaTraversalControl.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        if (control == Control.INSERT_AFTER) {
+            TreeTransformerUtil.insertAfter(context, element);
+        }
+        if (control == Control.INSERT_BEFORE) {
+            TreeTransformerUtil.insertAfter(context, element);
+        }
+        return TraversalControl.CONTINUE;
+    }
+}
```
Comment (bbakerman) [inline src/main/java/graphql/schema/visitor/GraphQLSchemaTraversalControl.java:81]: See how this hodes the TreeTransformerUtil malarky AND also we can detect intention from the GraphQLSchemaTraversalControl as a value and not a side effect

**Pull request merged**